tools/exact-pcap-match.c

```diff
@@ -300,177 +231,97 @@ int main (int argc, char** argv)
         }
 
-        pcap_pkthdr_t pkt_hdr;
-        if (read_expect (fd_ref, &pkt_hdr, sizeof(pkt_hdr), &offset, false))
-        {
-            ch_log_error(
-                    "Could not read enough bytes from %s at offset %li, (%li required)\n",
-                    options.input, offset, sizeof(pkt_hdr));
+        pkt_info_t pkt_info;
+        pkt_info = pcap_buff_next_packet(&ref_buff);
+        pcap_pkthdr_t* ref_hdr = ref_buff.hdr;
+        switch(pkt_info){
+        case PKT_EOF:
+            goto ref_pcap_loaded;
+        case PKT_OVER_SNAPLEN:
+            ch_log_fatal("Packet with index %d does not comply with snaplen: %d (data len is %d)\n", pkt_num, ref_buff.snaplen, ref_buff.hdr->len);
+        case PKT_LEN_TRUNCATED:
+            ch_log_warn("Packet data length (%d) is < capture length (%d).\n", ref_hdr->len, ref_hdr->caplen);
             break;
-        }
-
-        if (pkt_hdr.caplen > fhdr_ref.snaplen || pkt_hdr.caplen < 64)
-        {
-            ch_log_error(
-                    "Error, packet length (%li) out of range [64,%li] %u offset=%li\n",
-                    pkt_hdr.caplen, fhdr_ref.snaplen, offset);
-        }
-
-        if (options.verbose && pkt_num >= options.offset_ref
-                && (pkt_hdr.len + sizeof(expcap_pktftr_t) < pkt_hdr.caplen))
-        {
-            ch_log_warn("Warning: packet len %li < capture len %li\n",
-                        pkt_hdr.len, pkt_hdr.caplen);
-        }
-
-        char pbuf[1024 * 64] = { 0 };
-        if (read_expect (fd_ref, &pbuf, pkt_hdr.caplen, &offset, false))
-        {
+        default:
             break;
         }
 
-        expcap_pktftr_t* pkt_ftr = (expcap_pktftr_t*) (pbuf + pkt_hdr.caplen
-                - sizeof(expcap_pktftr_t));
-
-        if (pkt_num < options.offset_ref)
-        {
+        if (pkt_num < options.offset_ref){
             //Skip over packets in the reference file
             continue;
         }
 
-        if (options.verbose)
-        {
+        if (options.verbose){
             dprintf (STDOUT_FILENO, "ref,");
-            dprint_packet (STDOUT_FILENO, expcap, &pkt_hdr, pkt_ftr, pbuf, true,
-                           true);
+            dprint_packet (STDOUT_FILENO, expcap, ref_hdr, ref_buff.ftr, ref_buff.pkt, true, true);
         }
 
         value_t val;
         bzero (&val, sizeof(val));
-        val.pkt_hdr = pkt_hdr;
-        if (expcap)
-        {
-            val.pkt_ftr = *pkt_ftr;
+        val.pkt_hdr = *ref_hdr;
+        if (expcap){
+            val.pkt_ftr = *ref_buff.ftr;
         }
 
-        const int64_t caplen =
-                expcap ?
-                        pkt_hdr.caplen - sizeof(expcap_pktftr_t) :
-                        pkt_hdr.caplen;
-
         /*Use the whole packet as the key, and the header as the value */
-        hash_map_push (hmap, pbuf, caplen, &val);
+        hash_map_push (hmap, ref_buff.pkt, ref_hdr->len, &val);
```

Comment:
This is a bug. Need to use caplen, not packet len, and need to take into account expcap ftr.